pipeline/artifactory/artifactory.go

```diff
@@ -0,0 +1,348 @@
+	MimeType          string               `json:"mimeType,omitempty"`
+	Size              string               `json:"size,omitempty"`
+	Checksums         artifactoryChecksums `json:"checksums,omitempty"`
+	OriginalChecksums artifactoryChecksums `json:"originalChecksums,omitempty"`
+	URI               string               `json:"uri,omitempty"`
+}
+
+// artifactoryChecksums reflects the checksums generated by
+// Artifactory
+type artifactoryChecksums struct {
+	SHA1   string `json:"sha1,omitempty"`
+	MD5    string `json:"md5,omitempty"`
+	SHA256 string `json:"sha256,omitempty"`
+}
+
+// Pipe for Artifactory
+type Pipe struct{}
+
+// Description of the pipe
+func (Pipe) Description() string {
+	return "Releasing to Artifactory"
+}
+
+// Run the pipe
+//
+// Docs: https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#ArtifactoryRESTAPI-Example-DeployinganArtifact
+func (Pipe) Run(ctx *context.Context) error {
+	instances := len(ctx.Config.Artifactories)
+	if instances == 0 {
+		return pipeline.Skip("artifactory section is not configured")
+	}
+
+	// Check if for every instance we have a the target,
+	// the username and a secret (password or api key)
+	// If not, we can skip this pipeline
+	for i := 0; i < instances; i++ {
+		if ctx.Config.Artifactories[i].Target == "" {
+			return pipeline.Skip(fmt.Sprintf("artifactory section is not configured properly (missing target in artifactory %d)", i))
+		}
+
+		if ctx.Config.Artifactories[i].Username == "" {
+			return pipeline.Skip(fmt.Sprintf("artifactory section is not configured properly (missing username in artifactory %d)", i))
+		}
+
+		envName := fmt.Sprintf("ARTIFACTORY_%d_SECRET", i)
+		if os.Getenv(envName) == "" {
+			return pipeline.Skip(fmt.Sprintf("missing secret for artifactory %d: %s", i, ctx.Config.Artifactories[i].Target))
+		}
+	}
+
+	return doRun(ctx)
+}
+
+func doRun(ctx *context.Context) error {
+	if !ctx.Publish {
+		return pipeline.Skip("--skip-publish is set")
+	}
+
+	// Loop over all builds, because we want to publish
+	// every build to Artifactory
+	for _, build := range ctx.Config.Builds {
+		if err := runPipeOnBuild(ctx, build); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// runPipeOnBuild runs the pipe for every configured build
+func runPipeOnBuild(ctx *context.Context, build config.Build) error {
+	sem := make(chan bool, ctx.Parallelism)
+	var g errgroup.Group
+
+	// Lets generate the build matrix, , because we want to publish
+	// every target to Artifactory
+	for _, target := range buildtarget.All(build) {
+		sem <- true
+		target := target
+		build := build
+		g.Go(func() error {
+			defer func() {
+				<-sem
+			}()
+
+			return doBuild(ctx, build, target)
+		})
+	}
+
+	return g.Wait()
+}
+
+// doBuild runs the pipe action of the current build and the current target
+// This is where the real action take place
+func doBuild(ctx *context.Context, build config.Build, target buildtarget.Target) (err error) {
```

Comment:
maybe this function should be named `upload` or something? It is not really building here

Reply:
You are right. Renamed it to `upload`.